nnvm/python/nnvm/testing/check_computation.py

```diff
@@ -0,0 +1,451 @@
+        A dict mapping input variable names to ranges or just a single range
+        (the same for all variables). Input values will be generated from
+        uniform distributions on these ranges. `head_grads` can also be
+        assigned a range this way.
+
+    exclude_targets : Set[str]
+        Skip compiling and running anything for these targets.
+
+    only_targets : Set[str]
+        Test only for those targets from `ctx_list()` that are also in this set.
+
+    numerical_grads : bool or "if_possible"
+        Whether to additionally check against numerically computed gradients. If 'if_possible' is
+        passed (which is the default) then it will try to create a gradient computation graph and
+        then check gradients numerically only if this graph can be created (i.e. if there are some
+        operations with unimplemented gradients, it will just issue a warning).
+
+    delta : float
+        A small value used for numerical gradient computation (usually called h in textbooks).
+
+    atol : float
+        Absolute tolerance.
+
+    rtol : float
+        Relative tolerance.
+
+    ng_atol : float
+        Absolute tolerance for numerical grad checking.
+
+    ng_rtol : float
+        Relative tolerance for numerical grad checking.
+
+    ng_max_error : float
+        Discard numerical partial derivatives whose estimated error is larger than this value.
+
+    ng_max_discarded_frac : float
+        Allow discarding no more than this fraction of partial derivatives.
+
+    dump_graph : bool
+        Dump the graph even on success.
+    """
+
+    if numerical_grads not in [False, True, 'if_possible']:
+        raise ValueError("numerical_grads must be a bool or 'if_possible', not {}"
+                         .format(numerical_grads))
+
+    input_vars = [x for x in symbol.list_input_variables()]
+    input_dict = {x.attr('name'): x for x in input_vars}
+
+    if grad_input_vars is None:
+        grad_input_vars = input_vars
+
+    shape = shape.copy() if shape else {}
+
+    grad_input_vars_real = []
+    for x in grad_input_vars:
+        if isinstance(x, tuple):
+            if len(x) != 2:
+                raise ValueError("Expected (var_name, shape), not {}".format(x))
+            x_name = x[0] if isinstance(x[0], str) else x[0].attr('name')
+            if x_name in shape:
+                if shape[x_name] != x[1]:
+                    raise ValueError("Can't decide which shape to use for {}: "
+                                     "{} from shape or {} from grad_input_vars"
+                                     .format(x_name, shape[x_name], x[1]))
+            shape[x_name] = x[1]
+            x = x[0]
+
+        grad_input_vars_real.append(input_dict[x] if isinstance(x, str) else x)
+
+    grad_input_vars = grad_input_vars_real
+
+    # Infer the output shape and dtype by creating a graph and running passes
+
+    forward_graph = nnvm.graph.create(symbol)
+
+    if dtype is not None:
```

Comment:
We need to set_dtype_inputs. just check dtype is not None.

Reply:
Sorry, I think I didn't understand this one.

Reply:
If dtype is None, graph_to_function will raise exception? Same thing will happen for shape.
How about setting dtype as optional but not None as default value, such as dtype="float32"? For shape argument we should force user to pass in value. Then the code logic can be simpler.

Reply:
Setting dtype='float32' as default would simplify a lot of tests a little bit. However, if some variables have their dtypes specified by attributes then either these attributes should be preferred or an error should be raised, so I'm not sure.

Reply:
OK. Maybe we can put the shape/dtype check right in this function, before infer shape/dtype, and check either input variable shape/dtype attributes are set or shape/dtype arguments are passed.

Reply:
Since shapes and dtypes of some inputs may be inferred from shapes and dtypes of other inputs, the proper solution would be to run inference passes, and only then check if some inputs are left without types. Then these inputs can be assigned the default type.